File: java/src/main/java/br/datastructures/stack/ArrayStack.java
package br.datastructures.stack;

import br.datastructures.exceptions.NoSuchItemException;

public class ArrayStack<I> extends AbstractStack<I> {
    private int top, capacity;
    private I items[];
    
    private ArrayStack(int capacity, int top) {
        this.capacity = capacity;
        this.top = top;
        items = (I[]) new Object[capacity];
    }
    
    public ArrayStack() {
        this(5, -1);
    }
    
    public ArrayStack(int initialSize) {
        this(initialSize, -1);
    }
    
    private I[] resize() {
        I[] newItems = (I[]) new Object[capacity * 2];
        System.arraycopy(items, 0, newItems, 0, capacity);
        capacity *= 2;
        return newItems;
    }
    
    @Override
    public void push(I item) {
        if(top == capacity-1)
            items = resize();
        
        top++;
        items[top] = item;
        quantity++;
    }
    
    @Override
    public I pop() throws NoSuchItemException {
        if(empty())
            throw new NoSuchItemException();
        
        I item = items[top--];
        quantity--;
        
        return item;
    }
    
    
    @Override
    public I getTop() throws NoSuchItemException {
        if(empty())
            throw new NoSuchItemException();
        
        return items[top];
    }
}

File: c/ArrayStack.c
#include <stdio.h>
#include <stdlib.h>

typedef int Item;

typedef struct {
    int top, capacity;
    Item *items;
    int quantity;
} ArrayStack;

ArrayStack* newArrayStack(int initialSize) {
    ArrayStack *stack = (ArrayStack*) malloc(sizeof(ArrayStack));

    if(stack == NULL)
        return NULL;

    stack->items = (Item*) malloc(initialSize * sizeof(Item));
    
    if(stack->items == NULL)
        return NULL;

    stack->capacity = initialSize;
    stack->top = -1;
    stack->quantity = 0;

    return stack;
}

int size(ArrayStack* stack) {
    return stack->quantity;
}

int empty(ArrayStack* stack) {
    return size(stack) == 0;
}

Item* resize(ArrayStack* stack) {
    stack->capacity *= 2;
    Item* newItems = (Item*) malloc(stack->capacity * sizeof(Item));
    
    for(int i = 0; i < size(stack); i++) 
        newItems[i] = stack->items[i];

    free(stack->items);

    return newItems;
}

void push(ArrayStack* stack, Item item) {
    if(stack->top == stack->capacity-1)
        stack->items = resize(stack);
    
    stack->top++;
    stack->items[stack->top] = item;

    stack->quantity++;
}

Item pop(ArrayStack* stack) {
    if(empty(stack))
        return NULL;
    
    Item item = stack->items[stack->top--];
    stack->quantity--;
    
    return item;
}

Item getTop(ArrayStack* stack) {
    if(empty(stack)) 
        return NULL;
    
    return stack->items[stack->top];
}